casper/src/it/scala/coop.rchain.casper/HashSetCasperSpecification.scala

```diff
@@ -0,0 +1,212 @@
+    Gen.const(RNode(i, s"v$i", deployed = false))
+  }
+
+  def genCommand(state: State): Gen[Command] = for {
+    idx: Int <- Gen.chooseNum(0, state.size - 1)
+    n = state(idx)
+    c <- Gen.frequency(
+      (2, genDeployOrPropose(n)),
+      (1, genReceive(n))
+    )
+  } yield c
+
+  def genDeployOrPropose(n: RNode): Gen[Command] = {
+    if(n.deployed) Gen.const(Propose(n))
+    else Gen.frequency(
+      (1, Gen.const(Deploy(n))),
+      (1, Gen.const(DeployAndPropose(n)))
+    )
+  }
+
+  def genReceive(n: RNode): Gen[Command] = Gen.const(Receive(n))
+
+  case class Deploy(node: RNode) extends Command {
+    type Result = List[String]
+
+    def run(sut: Sut): Result = {
+      val v = sut(node.idx)
+      val rbm = (deploy(v.node, deployment(0)) >> create(v.node)).value.unsafeRunSync.right.get
+      v.update(Some(rbm))
+      v.node.logEff.errors ++ v.node.logEff.warns
+    }
+
+    def nextState(state: State): State = state.patch(node.idx, Seq(node.copy(deployed = true)), 1)
+
+    def preCondition(state: State): Boolean = !state(node.idx).deployed
+
+    def postCondition(state: State, result: Try[Result]): Prop = result.isSuccess && result.get.isEmpty
+
+  }
+
+  case class Propose(node: RNode) extends Command {
+    type Result = (BlockStatus, List[String])
+
+    def run(sut: Sut): Result = {
+      val v = sut(node.idx)
+      val r = add(v.node, v.lastBlock.get).value.unsafeRunSync.right.get.right.get
+      v.update(None)
+      (r, v.node.logEff.errors ++ v.node.logEff.warns)
+    }
+
+    def nextState(state: State): State =
+      state.patch(node.idx, Seq(node.copy(deployed = false)), 1)
+
+    def preCondition(state: State): Boolean = state(node.idx).deployed
+
+    def postCondition(state: State, result: Try[Result]): Prop = result.isSuccess && result.get._1 == Valid && result.get._2.isEmpty
+
+  }
+
+  case class DeployAndPropose(node: RNode) extends Command {
+    type Result = (BlockStatus, List[String])
+
+    def run(sut: Sut): Result = {
+      val v = sut(node.idx).node
+      val r = (deploy(v, deployment(0)) >> create(v) >>= (b => add(v, b))).value.unsafeRunSync.right.get.right.get
+      (r, v.logEff.errors ++ v.logEff.warns)
+    }
+
+    def nextState(state: State): State = state
+
+    def preCondition(state: State): Boolean = !state(node.idx).deployed
+
+    def postCondition(state: State, result: Try[Result]): Prop =
+      result.isSuccess && result.get._1 == Valid && result.get._2.isEmpty
+  }
+
+  case class Receive(node: RNode) extends Command {
+    override type Result = List[String]
+
+    def run(sut: Sut): List[String] = {
+      val v = sut(node.idx).node
+      v.receive().value.unsafeRunSync.right.get
+      v.logEff.errors ++ v.logEff.warns
+    }
+
+    def nextState(state: State): State = state
+
+    def preCondition(state: State): Boolean = true
+
+    def postCondition(state: State, result: Try[Result]): Prop = result.isSuccess && result.get.isEmpty
+  }
+}
```

Comment:
it would be great if you added comments.

Reply:
Comments tend to get outdated really quickly. Maybe some portions of code could be more readable?

Reply:
I meant comments describing the purpose of this tool: what is it used for, what it generated, how to continue etc. Not comments describing what the code does.